pkg/controllers/gateway/gateway_controller.go

```diff
@@ -351,6 +375,96 @@ func selectClusters(gateway gatewayv1beta1.Gateway) []string {
 	return []string{}
 }
 
+func (r *GatewayReconciler) DNSDeletion(ctx context.Context, gateway gatewayv1beta1.Gateway) error {
+	trafficAccessor := traffic.NewGateway(&gateway)
+	allHosts := trafficAccessor.GetHosts()
+
+	for _, host := range allHosts {
+		managedZone, subDomain, err := r.Host.GetManagedZoneForHost(ctx, host, trafficAccessor)
+		if err != nil {
+			return err
+		}
+		if managedZone == nil {
+			log.Log.Info("Managed zone not found cannot get DNS record")
+			return nil
+		}
+
+		log.Log.Info("getting DNS record associated with deleting gateway")
+		dnsRecord, err := r.Host.GetDNSRecord(ctx, subDomain, managedZone)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+		if dnsRecord == nil {
+			log.Log.Info("DNS record not found", "Host name:", host)
+			continue
+		}
+		err = r.Delete(ctx, dnsRecord)
+		log.Log.Info("DNS record deleted ", "DNSRecord:", dnsRecord.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *GatewayReconciler) cleanupCertificates(ctx context.Context, gateway *gatewayv1beta1.Gateway) error {
```

Comment:
same for this I think it could becomes part of  ```CertificateService``` WDYT?

Reply:
added booth deletion functions to interfaces. Hope I got the idea about "ingress gateway" correct :thinking: